crates/starfish/core/src/cordial_knowledge.rs

```diff
@@ -0,0 +1,1403 @@
+        for acknowledgment in header.acknowledgments() {
+            vec_knowledge_msgs[block_author].push(ConnectionKnowledgeMessage::RemoveShard {
+                block_ref: *acknowledgment,
+            });
+        }
+
+        // === 4) Traversing back and marking the causal past as known by block_author
+        // ===
+        let mut stack = vec![block_ref];
+        while let Some(current_ref) = stack.pop() {
+            let current_author = current_ref.author.value();
+            let current_round = current_ref.round;
+            let current_digest = current_ref.digest;
+
+            // ---- Get parents of current block ----
+            let parents_buf: Ancestors = {
+                let author_map = &self.cordial_knowledge[current_author];
+                let Some(current_round_map) = author_map.get(&current_round) else {
+                    continue;
+                };
+                let Some((parents, _)) = current_round_map.get(&current_digest) else {
+                    continue;
+                };
+                parents.clone()
+            };
+
+            // Traverse parents
+            for parent_ref in parents_buf.iter() {
+                let parent_author = parent_ref.author.value();
+                let parent_round = parent_ref.round;
+                let parent_digest = parent_ref.digest;
+
+                let parent_author_map = &mut self.cordial_knowledge[parent_author];
+
+                if let Some(parent_round_map) = parent_author_map.get_mut(&parent_round) {
+                    if let Some((_, who_knows_parent)) = parent_round_map.get_mut(&parent_digest) {
+                        // Mark that block_author now knows this parent
+                        if who_knows_parent.insert(block_author) {
+                            vec_knowledge_msgs[block_author].push(
+                                ConnectionKnowledgeMessage::RemoveHeader {
+                                    block_ref: *parent_ref,
+                                },
+                            );
+                            stack.push(*parent_ref);
+                        }
+                    }
+                }
+            }
+        }
+
+        // === 5) Send all accumulated knowledge messages ===
+        self.send_connection_knowledge_messages(vec_knowledge_msgs)
+            .await;
+    }
+
+    /// Send accumulated connection knowledge messages to all connection tasks.
+    async fn send_connection_knowledge_messages(&self, msgs: Vec<Vec<ConnectionKnowledgeMessage>>) {
+        for (index, msg) in msgs.into_iter().enumerate() {
+            if !msg.is_empty() {
+                let _ = self.connections[index].send(msg).await;
+            }
+        }
+    }
+
+    /// Report from Authority Service useful information about headers and
+    /// shards to global knowledge and connection knowledge.
+    pub async fn report_useful_info(
+        connection_knowledge_sender: &Sender<Vec<ConnectionKnowledgeMessage>>,
```

Comment:
`CordialKnowledge` already has all the `connection_knowledge_senders` in it's internal structure, perhaps we can only pass the peer index here, and retrieve the sender from the internal structure.